[Fix] Fix AttributeError in Qwen2.5 LoRA: 'Qwen2ForCausalLM' object has no attribute 'get_hidden_dim'

mssongit opened the pull request:
## Motivation

This PR fixes an `AttributeError` that occurs when using LoRA with the Qwen2.5 model. The error stems from an attempt to call `get_module_name` and `get_hidden_dim` on the `Qwen2ForCausalLM` model, which does not have this attribute. This breaks the LoRA initialization process, specifically when preallocating memory for the LoRA adapters.

## Modifications

Add fallback function for `get_module_name` and `get_hidden_dim`

## Checklist

- [x] Format the code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

mssongit committed:
init_lora_memory_pool hidden_dim update

AttributeError: 'Qwen2ForCausalLM' object has no attribute 'get_hidden_dim'

merrymercy commented:
@mssongit Could you fix the lint error?

mssongit committed:
Fix AttributeError in Qwen2.5(huggingface model) LoRA: 'Qwen2ForCausalLM' object has no attribute 'get_hidden_dim'

- Added a check for the existence of `get_hidden_dim` in `base_model`.
- Fallback to `base_model.config.hidden_size` if `get_hidden_dim` is not available.
- This ensures backward compatibility and prevents potential errors in models that don't implement `get_hidden_dim`.

mssongit commented:
> @mssongit Could you fix the lint error?

Sure! I’ve already fixed the lint error by ensuring the code properly checks for the existence of the get_hidden_dim method and falls back to config.hidden_size when necessary. I've also run the linter (black and others), and the issue should now be resolved. Let me know if you need any further adjustments or checks!

mssongit committed:
Fix AttributeError in Qwen2.5(huggingface model) LoRA: 'Qwen2ForCausalLM' object has no attribute 'get_module_name'

### Description
- Applied `black` formatting to `lora_manager.py` to ensure consistent code style and pass pre-commit linting checks.
- No functional changes were made, only code style adjustments for compliance with `black` standards.

### Changes
- Reformatted the `lora_manager.py` file using `black`.
- Ensured that the code adheres to the project's style guidelines.

### Testing
- Ran `pre-commit` hooks to ensure `black` formatting and other linting checks pass successfully.

mssongit committed:
Performance Improvement for LoRA Memory Pool Initialization

Optimized get_hidden_dim Calls:
The get_hidden_dim function is now called only once per module, and the result is reused to avoid redundant calculations, leading to better performance.
CUDA Memory Allocation Optimization:

Memory allocation using torch.empty has been optimized by checking if module_A and module_B already exist in A_buffer and B_buffer, respectively, to prevent redundant allocations.
Removal of Redundant Code:

Refactored redundant memory allocation and get_stacked_multiply calls, improving code readability and maintainability.

Ying1123 commented:
Hi @mssongit, thanks for the contribution! Could you provide a lora adapter path for testing?

Ying1123 commented on python/sglang/srt/lora/lora_manager.py:
```diff
@@ -104,12 +104,18 @@ def init_loras(self):
+            self.target_modules = {
+                name
+                for name, module in named_modules_dict.items()
+                if module in self.origin_target_modules
```
This does not seem correct. `self.origin_target_modules` is a list of strings, not modules. You can refer to the `get_module_name` function in `llama.py`.

mssongit replied:
Fix module name comparison for origin_target_modules in base model. This fix resolves errors encountered in models like Qwen2.5-14B Instruct LoRA, which do not have the get_module_name function.

Ying1123 replied:
Updated.

mssongit commented:
> Hi @mssongit, thanks for the contribution! Could you provide a lora adapter path for testing?

The base model I am using is Qwen/Qwen2.5-14B-Instruct, and the LoRA adapter has been uploaded to mssongit/Qwen2.5-14B-SFT-LoRA. Feel free to use that for testing!

mssongit committed:
Fix module name comparison for origin_target_modules in base model

Modified the logic to correctly handle self.origin_target_modules, which is a list of strings (module names), not actual module objects.
Updated the self.target_modules initialization:
If the base_model has a get_module_name method, use it to map origin_target_modules.
Otherwise, compare the module names in self.origin_target_modules with the names retrieved from named_modules() to ensure correct behavior.
This fix resolves errors encountered in models like Qwen2.5 LoRA, which do not have the get_module_name function.

mssongit committed:
Fix module name comparison for origin_target_modules in base model

Ying1123 committed:
fix the fallback solution of get_module_name

Ying1123 commented:
Hi @mssongit, I have directly modified your PR so that you may take a look. Currently, there is some issue with the base model "Qwen/Qwen2.5-14B-Instruct", which I'm still investigating. The fix for the base model will come up in another PR.
